File: lib/core/date_range.dart
/// Time period used by the dashboard aggregations.
enum Period { weekly, monthly, yearly }

extension PeriodX on Period {
  String get label {
    switch (this) {
      case Period.weekly:
        return 'Weekly';
      case Period.monthly:
        return 'Monthly';
      case Period.yearly:
        return 'Yearly';
    }
  }
}

class DateRange {
  final DateTime start; // inclusive
  final DateTime end; // exclusive
  const DateRange(this.start, this.end);

  bool contains(DateTime d) =>
      !d.isBefore(start) && d.isBefore(end);
}

class Periods {
  /// Returns the range for [period] containing [anchor].
  /// Weeks start on Monday.
  static DateRange rangeFor(Period period, DateTime anchor) {
    switch (period) {
      case Period.weekly:
        final monday = DateTime(anchor.year, anchor.month, anchor.day)
            .subtract(Duration(days: anchor.weekday - 1));
        return DateRange(monday, monday.add(const Duration(days: 7)));
      case Period.monthly:
        final start = DateTime(anchor.year, anchor.month, 1);
        final end = DateTime(anchor.year, anchor.month + 1, 1);
        return DateRange(start, end);
      case Period.yearly:
        return DateRange(
          DateTime(anchor.year, 1, 1),
          DateTime(anchor.year + 1, 1, 1),
        );
    }
  }

  /// Shifts [anchor] by [delta] periods (e.g. previous month).
  static DateTime shift(Period period, DateTime anchor, int delta) {
    switch (period) {
      case Period.weekly:
        return anchor.add(Duration(days: 7 * delta));
      case Period.monthly:
        return DateTime(anchor.year, anchor.month + delta, anchor.day);
      case Period.yearly:
        return DateTime(anchor.year + delta, anchor.month, anchor.day);
    }
  }

  static String title(Period period, DateRange range) {
    switch (period) {
      case Period.weekly:
        final endIncl = range.end.subtract(const Duration(days: 1));
        return '${range.start.day}/${range.start.month} - ${endIncl.day}/${endIncl.month}';
      case Period.monthly:
        const months = [
          'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
          'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
        ];
        return '${months[range.start.month - 1]} ${range.start.year}';
      case Period.yearly:
        return '${range.start.year}';
    }
  }
}

File: expense-manager-react/src/lib/date-range.ts
import { startOfWeek, endOfWeek, startOfMonth, endOfMonth, startOfYear, endOfYear, addWeeks, addMonths, addYears, subWeeks, subMonths, subYears, format } from 'date-fns';

export enum Period {
  Weekly = 'weekly',
  Monthly = 'monthly',
  Yearly = 'yearly',
}

export interface DateRange {
  start: Date;
  end: Date;
}

export function rangeFor(period: Period, anchor: Date): DateRange {
  switch (period) {
    case Period.Weekly:
      return {
        start: startOfWeek(anchor, { weekStartsOn: 1 }),
        end: endOfWeek(anchor, { weekStartsOn: 1 }),
      };
    case Period.Monthly:
      return {
        start: startOfMonth(anchor),
        end: endOfMonth(anchor),
      };
    case Period.Yearly:
      return {
        start: startOfYear(anchor),
        end: endOfYear(anchor),
      };
  }
}

export function shift(period: Period, anchor: Date, delta: number): Date {
  switch (period) {
    case Period.Weekly:
      return delta > 0 ? addWeeks(anchor, delta) : subWeeks(anchor, -delta);
    case Period.Monthly:
      return delta > 0 ? addMonths(anchor, delta) : subMonths(anchor, -delta);
    case Period.Yearly:
      return delta > 0 ? addYears(anchor, delta) : subYears(anchor, -delta);
  }
}

export function title(period: Period, range: DateRange): string {
  switch (period) {
    case Period.Weekly:
      return `${format(range.start, 'MMM d')} – ${format(range.end, 'MMM d, yyyy')}`;
    case Period.Monthly:
      return format(range.start, 'MMMM yyyy');
    case Period.Yearly:
      return format(range.start, 'yyyy');
  }
}

export function getPeriodUnits(period: Period, range: DateRange): string[] {
  switch (period) {
    case Period.Weekly:
      return ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    case Period.Monthly: {
      const weeks: string[] = [];
      const start = range.start;
      const end = range.end;
      let current = start;
      let weekNum = 1;
      while (current <= end) {
        weeks.push(`W${weekNum}`);
        weekNum++;
        current = addWeeks(current, 1);
      }
      return weeks;
    }
    case Period.Yearly:
      return ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  }
}
